### ISSUE TITLE
echo and \0

### ISSUE THREAD

**Issue by dlegaultbbry** [landley/toybox#193]
I just integrated the latest version (0.8.3) and I get test failures in strings but it seems related to changes in echo.

Looks like echo doesn't handle \0 right anymore

```
FAIL: strings string and nulls
/data/toybox/test/xbin/echo -ne "\0foobar\0\0bazfoo\0foo" > input
/data/toybox/test/xbin/echo -ne '' | strings input
--- expected	2020-05-20 15:12:54.000000000 +0000
+++ actual	2020-05-20 15:12:54.000000000 +0000
@@ -1,2 +1 @@
-foobar
-bazfoo
+\0foobar\0\0bazfoo\0foo
```

The unescape2 function seems to fail for this case.

echo also doesn't have a \0 test so perhaps the issue slipped through unnoticed.

**Comment by enh-google** [comment]
i was going to say "oddly, the test passes on Android", but that appears to be because we're getting the mksh echo:
```
vsoc_x86_64:/ # echo -ne "\0foobar\0\0bazfoo\0foo"  | xxd                                                                  
00000000: 0066 6f6f 6261 7200 0062 617a 666f 6f00  .foobar..bazfoo.
00000010: 666f 6f                                  foo
vsoc_x86_64:/ # /bin/echo -ne "\0foobar\0\0bazfoo\0foo"  | xxd                                                             
00000000: 5c30 666f 6f62 6172 5c30 5c30 6261 7a66  \0foobar\0\0bazf
00000010: 6f6f 5c30 666f 6f                        oo\0foo
vsoc_x86_64:/ # 
```

**Comment by dlegaultbbry** [comment]
Same here, the shell we use (ksh) has a builtin echo so everywhere in the test framework where echo is found, I had to convert to $(which echo) to make sure it picked up the right one from toybox

**Comment by dlegaultbbry** [comment]
Also breaks the env -0 test.

**Comment by landley** [comment]
I don't have a tests/xbin.test here?

Android is bypassing half my test plumbing: testcmd calls $C instead of $CMDNAME precisely so it _doesn't_ use shell builtins, and do_test() sets $C to the path to the absolute path to the command, either in $TESTDIR or via $(which $CMDNAME) as appropriate for $TEST_HOST.

I did that for a reason. Not doing that is why you're testing shell builtins.

**Comment by landley** [comment]
Try commit 310eefe29567

**Comment by dlegaultbbry** [comment]
It fixes stuff when \0 is in the middle of text (strings test now passes fine), but I still see a failure on the env -0 test. There's an extra \0 printed at the end of the expected result.

FAIL: env -0
/data/toybox/test/xbin/echo -ne '' | "/data/toybox/test/xbin/env" -i five=six seven=eight /data/toybox/test/xbin/env -0 | sort -z
--- expected	2020-05-21 10:31:07.000000000 +0000
+++ actual	2020-05-21 10:31:07.000000000 +0000
@@ -1 +1 @@
-five=sixseven=eight\
\ No newline at end of file
+five=sixseven=eight
\ No newline at end of file

echo -ne "five=six\0seven=eight\0"
five=sixseven=eight\0

expected is 

five=sixseven=eight

**Comment by enh-google** [comment]
> I don't have a tests/xbin.test here?
> 
> Android is bypassing half my test plumbing: testcmd calls $C instead of $CMDNAME precisely so it _doesn't_ use shell builtins, and do_test() sets $C to the path to the absolute path to the command, either in $TESTDIR or via $(which $CMDNAME) as appropriate for $TEST_HOST.
> 
> I did that for a reason. Not doing that is why you're testing shell builtins.

the change that made $C be like this (https://android-review.git.corp.google.com/c/platform/external/toybox/+/1007687) claims to have been to avoid running the builtin kill.

i think the issue here is that this isn't the _echo_ test (where we do explicitly use the toybox echo), it's the _strings_ test. so we're not doing anything special to ensure we get the toybox echo.

unlike regular host testing we [mostly; there are exceptions like blkid] don't have to worry about the "host" non-toybox executable. so it's only the handful of shell builtins when they're used in _other_ tests. not sure what, if anything, to do about that.

**Comment by dlegaultbbry** [comment]
Analyzed stuff a bit more and there's a bug for sscanf in unescape2 and because of this, the off jumps past the end of the string and continues printing junk until it hits another real string terminator.

The check needs to be changed to if (sscanf(...) > 0) as a an ending \0 will attempt to trigger on the octal string, but it returns EOF because it didn't match before reaching end of input. Because the return value is not checked for errors, the EOF (-1) triggers the if and we jump past the end of the string and hell ensues.

**Comment by enh-google** [comment]
patch with test sent to list: http://lists.landley.net/pipermail/toybox-landley.net/2020-August/011908.html

**Comment by enh-google** [comment]
fixed by 310eefe2956725ab1ae913a2eccf3f885e321c7f and 3ba988c453056cbdda2bc26f1b9cbc3b4697acd0.
